### DISCUSSION TITLE
SwiftData를 통해 오늘 날짜에 해당하는 값을 받아오고 싶습니다.

### DISCUSSION THREAD

**Discussion by sseungwonnn** [DeveloperAcademy-POSTECH/.github#33] [Q&A] [answered]
### 😀 야호
안녕하세요. 오후반 슌입니다.
MC2 때 일어났던 문제를 이제서야 공유하네요. 첫 공유! 야호~

SwiftData를 통해 데이터를 저장했었고 '오늘' 날짜에 해당하는 값을 불러오고 싶었으나 잘 되지 않았습니다.
방법을 찾고 싶습니다 ~~

먼저 해당 코드 첨부하겠습니다.
``` swift
    func fetchTodayMeal() -> [MealData] {
        var meals: [MealData] = []
        
        let predicate = #Predicate<MealData> { $0.date == Date.now }
        
        let descriptor = FetchDescriptor(predicate: predicate)

        do {
            let storedMeals = try modelContext.fetch(descriptor)
            print(storedMeals)
            meals = storedMeals
        } catch {
            print("\(MealDataError.notFound)")
        }

        return meals
    }
```

---

### ⚠️ 문제
<img width="1512" alt="image" src="https://github.com/DeveloperAcademy-POSTECH/.github/assets/74140181/be09f624-3d4a-401e-a8b5-527bfc7b9da2">
사진과 같이 "Key path cannot refer to static member 'now'" 라는 에러가 발생합니다.
말 그대로, 현재 날짜를 불러오는 Date.now에서 static member를 참조할 수 없어 보입니다.

---

### ‼️ 시도
1. 현재처럼 SwiftDataManager 클래스를 만들어서 predicate를 활용해서 오늘 날짜 데이터를 불러오기
**-> 실패**
2. 해당하는 값이 필요한 뷰에서 ```@Query``` Annotation을 통해 불러오기
**-> 실패**

해당 코드
```swift
    // 방법 2-1
    @Query(filter: #Predicate<MealData> { mealItem in
        mealItem.date == Date.now
    }) var todayMeals: [MealData]

    // 방법 2-2
    @Query(filter: #Predicate<MealData> { mealItem in
        Calendar.current.startOfDay(for: mealItem.date) == Calendar.current.startOfDay(for: Date())
    }) var todayMeals: [MealData]
```
마찬가지로 같은 에러가 발생합니다.

3. 결국 타협한 점, 전체 데이터를 ```@Query``` Annotation을 통해 불러오고, Computed properties 를 통해 값을 초기화하기
**-> 성공했으나 찜찜..**

해당 코드
```swift
    // 모든 식사
    @Query var everyMeal: [MealData]
    
    // 오늘의 식사
    var todayMeals: [MealData] {
        return everyMeal.filter({$0.date == Date.now})
    }
```

모든 값들을 불러온 다음에 해당 날짜만 뽑아내는 게 너무 비효율적인 것 같습니다. 10000개를 불러오고 3개만 뽑아내기 너무 아쉬워요.

---

4. *추가된 내용: Date extension을 통해 dateFormatter를 만들어, 날짜만을 String 타입으로 받아와서 비교해보았습니다.

해당 코드

```swift
    // 모든 식사
     @Query(filter: #Predicate<MealData> { mealItem in
        mealItem.date.dateFormatter == Date.now.dateFormatter
    }) var todayMeals: [MealData]
```

String 타입으로도 비교를 했으나 역시 1, 2번과 마찬가지로 같은 에러가 발생했습니다.

---

### ⁉️ 궁금한 점
- 3번 방법 말고 다른 방법으로 오늘 날짜인 데이터를 저장소에서 불러올 수 있는지 궁금합니다!
- 혹은 아예 다른 방식은 무엇이 있을까요?

**Comment by YunKi-H** [comment]
2번 방법에서 혹시 실패한 이유가 뭔가요?
Date.now 는 현재 시점의 Date 라서 저장된 date 와 비교하면 반드시 false 가 반환될 것 같습니다.
ex) 2024.06.28 00:00:00 != 2024.06.28 08:55:13

**Reply by sseungwonnn** [reply]
마찬가지로, predicate 안에 static member인 .now가 들어갈 수 없다고 나오네요. 

또한 본문 업데이트 했습니다. 시간은 없애고 날짜만을 비교하도록 했는데도 같은 문제가 발생하네요😢

**Reply by sseungwonnn** [reply]
\* 내용 추가합니다.
MealData에는 date.dateFormatter 라는 Property 가 없다고 에러를 반환하네요. 
date 까지 인식하는 것이 아닌, date.dateFormatter 까지를 하나로 보는 것 같습니다.

**Comment by thinkySide** [comment, marked as answer]
팀즈에 올린 내용 도움이 되실까 싶어 Github에도 공유드립니다!

저도 이전에 같은 문제를 겪고 있다 해결해 공유드립니다!
Date.now를 Predicate 구문 밖에 변수로 만들고 시도해보시면 될 것 같아요!

~~~swift
let now = Date.now
let predicate = #Predicate<MealData> { $0.date == now }
~~~

저도 정확한 이유는 알 수없지만,,, 추측을 해보자면! (정말 개인적인 추측입니다)
Predicate 매크로 안에서 Date.now는 정확히 특정할 수 없기 때문에?
그렇지 않을까,,, (매크로 동작 방식에 대한 이해가 더 필요할 것 같기도 합니다.) 생각하고 있습니다!

**Reply by sseungwonnn** [reply]
답변 감사합니다 !

Predicate 구문 밖에 변수로 만들고 실행해보았는데, 
1. 에러는 사라져서 빌드는 성공적으로 됩니다.
2. 하지만 값이 불러와지지는 않네요...

```swift
func fetchTodayMeal() -> [MealData] {
        var meals: [MealData] = []
        
        let now = Date.now
        let predicate = #Predicate<MealData> { $0.date == now }
        
        let descriptor = FetchDescriptor(predicate: predicate)
        
        do {
            let storedMeals = try modelContext.fetch(descriptor)
            print(storedMeals)
            meals = storedMeals
        } catch {
            print("\(MealDataError.notFound)")
        }
        
        return meals
    }
```

해당 값이 필요한 부분에서는 다음과 같이 초기화해서 사용했습니다.
```swift
    var todayMeals: [MealData] {
        return swiftDataManger.fetchTodayMeal()
    }
```

하지만 여전히 본문에 실린 방법 3의 코드는 동작합니다.
```swift
    @Query var everyMeal: [MealData]
    var todayMeals: [MealData] {
        return everyMeal.filter({$0.date == Date.now })
    }
```

**Comment by Jaehwa-Noh** [comment]
안녕하세요. 1기 제다이 노재화입니다.

Swift의 한계로 보입니다.  Static member를 key path로 사용할 수 없다네요.
https://forums.swift.org/t/key-path-cannot-refer-to-static-member/28055

@thinkySide 님이 말씀하신대로 하셔야할거 같아보이네요.
그나저나 Xcode도 많이 발전했군요. 내부 오류 코드도 보여주고.

**Reply by sseungwonnn** [reply]
답변 감사합니다!

첨부해주신 링크 잘 읽었습니다.
static 을 활용하지 않고 참조할 수 있는 다른 방법을 찾아봐야겠네요..

**Reply by Jaehwa-Noh** [reply]
로직을 몰라서 궁금한 게 있는데요.

Date.now는 시:분:초가 계속해서 현재 시각으로 변경되는데 완전히 일치하는 게 존재할 수 있나요?
> $0.date == Date.now

또, `MealData`의 구조는 어떻게 되어있나요?

**Reply by sseungwonnn** [reply]
존재할 수 없는 것 같습니다. 답변해주신 내용을 살펴보니, MealData를 초기화할 때, 시:분:초 를 제거하고 날짜만을 담고, 이후에 ```$0.date == now``` 를 사용하면 해결할 수 있을 것 같습니다.

추가로 MealData 구조 첨부합니다.
```swift
@Model
class MealData: Equatable, Hashable {
    var date: Date
    var mealTime: String // 아침, 점심, 저녁
    
    var preMealBloodSugarLevel: Int?
    var postMealBloodSugarLevel: Int?
    
    var menu: String?
    var mealSize: String? // 적음, 보통, 많음
    var memo: String?
    
    var calculatedHighInsulinDosage: Double?
    var calculatedLowInsulinDosage: Double?
    
    var insulinDosage: Int?
    var increasedBloodSugarLevel: Double?
   
    // init 생략
}
```

**Comment by DevKang** [comment]
안녕하세요 :-) 아직 해결 중이신지는 모르겠는데, 한번 참고 해주세요 👍

<img width="459" alt="image" src="https://github.com/DeveloperAcademy-POSTECH/.github/assets/2010236/d3cbe3bf-2430-429c-a4c9-7a344d65ad87">

1. 이미 해결하신 이슈이지만, 
이 Predicate는 내가 만든 레코드의 Date 타입의 `date` 속성 값이 현재 시각을 나타내는 `Date.now` 와 "정확히 같은지" 를 물어보는 구문입니다.
Date 타입은 기본 연산 단위가 ms 이기 때문에, 두 `Date` 타입의 값이 같은지 연산하게 되면 정확하게 일치하는 값이 없을 것이기 때문에 아무 값도 가져오지 못하게 됩니다.

해결하신 것 처럼 날짜만을 가져와서 계산 하는 방법도 있겠지만, 부등호 연산자를 활용하는 방법도 있어서 함께 공유드려요. 아래 코드는 제가 작업하고 있는 다른 앱에서 발췌한 코드입니다.

```swift
let today = Date.now.startOfDay ?? .now
_todayComments = Query(
    filter: #Predicate { $0.created > today },
    sort: [SortDescriptor(\.created, order: .reverse)]
)

// 위의 코드에서 사용된 startOfDay 에 대한 extension

extension Date {
    // 오늘의 00시를 반환하는 계산된 프로퍼티
    var startOfDay: Date? {
        let calendar = Calendar.current
        // 오늘의 00시를 나타내는 DateComponents 생성
        var components = calendar.dateComponents([.year, .month, .day], from: self)
        components.hour = 0
        components.minute = 0
        components.second = 0
        // 오늘의 00시를 나타내는 Date 객체 생성
        return calendar.date(from: components)
    }
}
```

2. Predicate를 만들 때 필요한 속성 값들은 미리 다 계산해서 변수 혹은 상수로 만들어 놓은 이후에 Predicate 구분은 단순 연산으로 표현해주셔야 합니다. 
위의 코드에서도 Predicate 구문 안에서 `Date.now.startOfDay ?? .now`를 사용한게 아니라 위에서 상수로 만든 이후에 사용해야 매크로도 잘 동작하고, 실제 SwiftData안에서 동작도 무리 없이 실행 됩니다.

두서 없이 작성하긴 했지만, 잘 이해하셨을 것이라 믿습니다 ㅋㅋ

**Reply by sseungwonnn** [reply]
답변 감사합니다~
부등호 연산으로 받아오는 건 처음 봤습니다 👍 적용해보겠습니다!
